### PR TITLE
Making TransitioningAttributeStorage "non-eager", throttling et al.

### PR DESCRIPTION
**Pull request by cstamas**
Attribute upgrade from now on happens only on attribute STORE operation,
while attribute READ operation just _reads_ them (from new or falls back to old
storage).

Just for review, not for vote yet. This change alone would not be sufficient for sure, coz some tests expect the "old" behavior.

### CODE REVIEW

**Comment by velo** [comment]
+1

**Comment by bdemers** [comment]
We should have tests for this
(not sure the best way to test throttling)
other then that it looks good

File: nexus/nexus-api/src/main/java/org/sonatype/nexus/proxy/walker/WalkerThrottleController.java [hunk 1]
```diff
@@ -0,0 +1,77 @@
+package org.sonatype.nexus.proxy.walker;
+
+/**
+ * Interface to controle throttling (if desired) in executing of Walks.
+ *
```
Comment (peterlynch) [inline nexus/nexus-api/src/main/java/org/sonatype/nexus/proxy/walker/WalkerThrottleController.java:5]: sp.

File: nexus/nexus-proxy/src/main/java/org/sonatype/nexus/proxy/attributes/upgrade/DefaultAttributeUpgrader.java [hunk 2]
```diff
@@ -0,0 +1,204 @@
...
+            getLogger().debug( "Legacy attribute directory not present, no need for attribute upgrade." );
+        }
+        else
+        {
+            if ( isUpgradeDone( getLegacyAttributesDirectory(), null ) )
+            {
+                // nag the user to remove the directory
+                getLogger().info(
+                    "Legacy attribute directory present, but is marked already as upgraded. Please delete, move or rename the \""
+                        + getLegacyAttributesDirectory().getAbsolutePath() + "\" folder." );
```
Comment (peterlynch) [inline nexus/nexus-proxy/src/main/java/org/sonatype/nexus/proxy/attributes/upgrade/DefaultAttributeUpgrader.java:132]: slf4j logger? could use {} templating

File: nexus/nexus-proxy/src/main/java/org/sonatype/nexus/proxy/attributes/upgrade/DefaultAttributeUpgrader.java [hunk 3]
```diff
@@ -0,0 +1,204 @@
...
+                final File target = new File( new File( attributesDirectory, repoId ), MARKER_FILENAME );
+                // this step is needed if new repo added while upgrade not done: it will NOT have legacy attributes
+                // as other reposes, that were present in old/upgraded instance
+                target.getParentFile().mkdirs();
+                FileUtils.fileWrite( target, MARKER_TEXT );
+            }
+        }
+        catch ( IOException e )
+        {
+            // hum?
```
Comment (peterlynch) [inline nexus/nexus-proxy/src/main/java/org/sonatype/nexus/proxy/attributes/upgrade/DefaultAttributeUpgrader.java:201]: finish that thought?

File: nexus/nexus-proxy/src/main/java/org/sonatype/nexus/proxy/attributes/upgrade/UpgraderThread.java [hunk 4]
```diff
@@ -0,0 +1,175 @@
...
+    @Override
+    public void run()
+    {
+        try
+        {
+            Thread.sleep( 5000 );
+        }
+        catch ( InterruptedException e )
+        {
+            return;
```
Comment (peterlynch) [inline nexus/nexus-proxy/src/main/java/org/sonatype/nexus/proxy/attributes/upgrade/UpgraderThread.java:76]: Shouldn't Thread.currentThread().interrupt(); be called here?

File: nexus/nexus-proxy/src/main/java/org/sonatype/nexus/proxy/attributes/upgrade/UpgraderThread.java [hunk 5]
```diff
@@ -0,0 +1,175 @@
...
+    {
+        this.limiterUps = limiterUps;
+    }
+
+    @Override
+    public void run()
+    {
+        try
+        {
+            Thread.sleep( 5000 );
```
Comment (peterlynch) [inline nexus/nexus-proxy/src/main/java/org/sonatype/nexus/proxy/attributes/upgrade/UpgraderThread.java:72]: nit: personally prefer to see

```
TimeUnit.SECONDS.sleep(5);
```

**Comment by cstamas** [comment]
Am closing this pull request for now, to separate it into two pulls, one about substantial fixes (that should go ASAP into master), and one about the "throttling" where we can continue this debate.

**Pull request closed without merging**